subscriptions/functions.py

```diff
@@ -280,3 +280,35 @@ def cache_clear(self_, *args, **kwargs) -> bool:
 def get_default_features() -> Set[Feature]:
     default_tiers = Tier.objects.filter(is_default=True).prefetch_related('features')
     return merge_feature_sets(*(tier.features.all() for tier in default_tiers))
+
+
+def get_resource_refresh_moments(
+    user: AbstractUser,
+    at: Optional[datetime] = None,
+    assume_subscription_refresh: bool = True,
+) -> dict[Resource, datetime]:
+    """
+    For given user and moment in time, provides information when all the resources will be refreshed.
+    If the given resource won't be refreshed (e.g. because subscription ends), it will be absent from the dictionary.
+
+    If `assume_subscription_refresh` is set to `True` we allow
+    the recharge moments to be beyond the current subscription end.
+    """
+    at = at or now()
+    result = {}
+    datetime_max = datetime.max.replace(tzinfo=timezone.utc)
+
+    for subscription in iter_subscriptions_involved(user, at):
```

Comment:
`iter_subscriptions_involved` is something more internal than for general usage; it returns all the subscriptions that are involved in resource calculations at `at` moment. It may include subscriptions which already ended but could potentially have some impact on final resource count. See `test_subscriptions_involved`.

If you want to get resource refresh information, you should use `iter_subscriptions_quota_chunks` since `now()` till `INFINITY` or so. It will yield all resources' quota updates. No need to calculate recharge moment manually.

If you want to assume subscription refresh, you can do something like this:
```python
subscriptions = user.subscriptions.active()
for subscription in subscriptions:
    # make subscription infinite unless the plan is hard-limited in time
    subscription.end = max(
        subscription.end,  # in case admin manually shifted subscription end
        subscription.start + (plan.max_duration or INFINITY)
    )
    # note: we do not save subscription
for quota_chunk in iter_subscriptions_quota_chunks(subscriptions, since=now(), until=INFINITY):
    # for each resource, pick the first quota chunk that starts after now() - that's earliest recharge for that resource
```